Use Mailgun Signing key instead of API Key

Matthijsy opened the pull request:
### Summary
Mailgun needs the signing key instead of the API key for validating incoming mails. For this reason the naming has been changed. See also issue #38471

### Other Information
This introduces a deprecation of a credential key and an environment variable. I could not find what the normal workflow for deprecation is, so I have included a deprecation warning and left the version in which this will be removed open. Please let me know if I have to do something different for this.

georgeclaghorn commented:
Does Mailgun still sign some webhook requests with the API key? Are they phasing signing keys in or something? Is there another reason to deprecate the API key instead of failing hard when we don’t have a signing key?

Matthijsy commented:
My main reason was to give users the possibility to migrate away from using it instead of having their application fail due to this change. It would be fine to do a hard fail but I don't know what the policies about backwards compatibility are?

When you create an initial account on mailgun the API key and signing key are equal (at least when I created an account) so this could mean that there are users which use their 'API key' (which is also their signing key) and have actionmailbox working.

georgeclaghorn commented:
> When you create an initial account on mailgun the API key and signing key are equal (at least when I created an account) so this could mean that there are users which use their 'API key' (which is also their signing key) and have actionmailbox working.

Just checked my testing account and the API key and signing key are indeed the same. +1 for deprecating, then.

eugeneius commented on actionmailbox/CHANGELOG.md:
```diff
@@ -1,3 +1,7 @@
+*   Deprecated `Rails.application.credentials.action_mailbox.api_key` and `MAILGUN_INGRESS_API_KEY` in favor of `Rails.application.credentials.action_mailbox.signing_key` and `MAILGUN_INGRESS_SIGNING_KEY`
```
Changelog entries are usually written in the imperative mood:

```suggestion
*   Deprecate `Rails.application.credentials.action_mailbox.api_key` and `MAILGUN_INGRESS_API_KEY` in favor of `Rails.application.credentials.action_mailbox.signing_key` and `MAILGUN_INGRESS_SIGNING_KEY`.
```

eugeneius commented on actionmailbox/app/controllers/action_mailbox/ingresses/mailgun/inbound_emails_controller.rb:
```diff
@@ -16,21 +16,21 @@ module ActionMailbox
   # == Usage
   #
-  # 1. Give Action Mailbox your {Mailgun API key}[https://help.mailgun.com/hc/en-us/articles/203380100-Where-can-I-find-my-API-key-and-SMTP-credentials-]
+  # 1. Give Action Mailbox your {Mailgun Signing key}[https://help.mailgun.com/hc/en-us/articles/203380100-Where-can-I-find-my-API-key-and-SMTP-credentials-]
```
This link points to documentation for finding an API key, not a signing key.

eugeneius commented on actionmailbox/app/controllers/action_mailbox/ingresses/mailgun/inbound_emails_controller.rb:
```diff
@@ -16,21 +16,21 @@ module ActionMailbox
   #
   #        action_mailbox:
-  #          mailgun_api_key: ...
+  #          mailgun_singing_key: ...
```
🎤🔑

eugeneius commented on actionmailbox/app/controllers/action_mailbox/ingresses/mailgun/inbound_emails_controller.rb:
```diff
@@ -64,14 +64,22 @@ def authenticated?
-        Rails.application.credentials.dig(:action_mailbox, :mailgun_api_key) || ENV["MAILGUN_INGRESS_API_KEY"]
+        if Rails.application.credentials.dig(:action_mailbox, :mailgun_api_key) ||  ENV["MAILGUN_INGRESS_API_KEY"]
+          ActiveSupport::Deprecation.warn(<<-MSG.squish)
+            Rails.application.credentials.action_mailbox.api_key and ENV.MAILGUN_INGRESS_API_KEY are deprecated and will be removed from Rails x.x.
```
This deprecation will be added to 6.1, so the message should mention 6.2:

```suggestion
            Rails.application.credentials.action_mailbox.api_key and ENV.MAILGUN_INGRESS_API_KEY are deprecated and will be removed from Rails 6.2.
```

eugeneius commented on actionmailbox/app/controllers/action_mailbox/ingresses/mailgun/inbound_emails_controller.rb:
```diff
@@ -64,14 +64,22 @@ def authenticated?
 
       def key
-        Rails.application.credentials.dig(:action_mailbox, :mailgun_api_key) || ENV["MAILGUN_INGRESS_API_KEY"]
+        if Rails.application.credentials.dig(:action_mailbox, :mailgun_api_key) ||  ENV["MAILGUN_INGRESS_API_KEY"]
```
We could be more specific about which deprecated configuration method is being used:

```ruby
if Rails.application.credentials.dig(:action_mailbox, :mailgun_api_key)
  ActiveSupport::Deprecation.warn(<<-MSG.squish)
    Rails.application.credentials.action_mailbox.api_key is deprecated and will be ignored in Rails 6.2.
    Use Rails.application.credentials.action_mailbox.signing_key instead.
  MSG
  Rails.application.credentials.dig(:action_mailbox, :mailgun_api_key)
elsif ENV["MAILGUN_INGRESS_API_KEY"]
  ActiveSupport::Deprecation.warn(<<-MSG.squish)
    The MAILGUN_INGRESS_API_KEY environment variable is deprecated and will be ignored in Rails 6.2.
    Use MAILGUN_INGRESS_SIGNING_KEY instead.
  MSG
  ENV["MAILGUN_INGRESS_API_KEY"]
else
  Rails.application.credentials.dig(:action_mailbox, :mailgun_signing_key) || ENV["MAILGUN_INGRESS_SIGNING_KEY"]
end
```

eugeneius reviewed (commented):
The Action Mailbox guide needs to be updated too: https://github.com/rails/rails/blob/master/guides/source/action_mailbox_basics.md

There's also two lines in these files that still mention the API key:
https://github.com/rails/rails/blob/1e7f867a7a14403d549edd6b67949d9598192e14/actionmailbox/app/controllers/action_mailbox/ingresses/mailgun/inbound_emails_controller.rb#L9
https://github.com/rails/rails/blob/1e7f867a7a14403d549edd6b67949d9598192e14/actionmailbox/test/controllers/ingresses/mailgun/inbound_emails_controller_test.rb#L70

eugeneius commented on guides/source/action_mailbox_basics.md:
```diff
@@ -76,19 +76,19 @@ bin/rails action_mailbox:ingress:exim URL=https://example.com/rails/action_mailb
+Mailgun Signing key (which you can find under Settings -> Security & Users -> API security in Mailgun)
 so it can authenticate requests to the Mailgun ingress.
 
 Use `bin/rails credentials:edit` to add your API key to your application's
```
```suggestion
Use `bin/rails credentials:edit` to add your Signing key to your application's
```

eugeneius commented on guides/source/action_mailbox_basics.md:
````diff
@@ -76,19 +76,19 @@ bin/rails action_mailbox:ingress:exim URL=https://example.com/rails/action_mailb
 ```yaml
 action_mailbox:
-  mailgun_api_key: ...
+  mailgun_singing_key: ...
````
```suggestion
  mailgun_signing_key: ...
```

eugeneius commented on guides/source/action_mailbox_basics.md:
````diff
@@ -76,19 +76,19 @@ bin/rails action_mailbox:ingress:exim URL=https://example.com/rails/action_mailb
 ```
 
-Alternatively, provide your API key in the `MAILGUN_INGRESS_API_KEY` environment
+Alternatively, provide your API key in the `MAILGUN_INGRESS_SIGNING_KEY` environment
````
```suggestion
Alternatively, provide your Signing key in the `MAILGUN_INGRESS_SIGNING_KEY` environment
```

Matthijsy committed:
Use Mailgun Singing key instead of API Key

Mailgun needs the signing key instead of the API key for validating incoming mails. For this reason the naming has been changed.

georgeclaghorn commented:
Thank you!
